build-logic/packaging/src/main/kotlin/gradlebuild/instrumentation/tasks/InstrumentedMetadataMergeTask.kt

```diff
@@ -144,6 +164,34 @@ abstract class FindInstrumentedSuperTypesTask : DefaultTask() {
         }
     }
 
+    private
+    fun mergeAndWriteUpgradedProperties() {
+        // Merge and find all upgraded properties
+        val mergedUpgradedProperties = mergeProperties()
+        if (mergedUpgradedProperties.isEmpty) {
+            upgradedProperties.asFile.get().toEmptyFile()
+            upgradedPropertiesHash.asFile.get().toEmptyFile()
+            return
+        }
+
+        upgradedProperties.asFile.get().writer().use { Gson().toJson(mergedUpgradedProperties, it) }
+        val hasher = Hashing.newHasher()
+        mergedUpgradedProperties.map { it.asJsonObject.get("hash").asString }.sorted().forEach { hasher.putString(it) }
+        upgradedPropertiesHash.asFile.get().outputStream().use { it.write(hasher.hash().toByteArray()) }
```

Comment:
❓ Is this hash required here at all? I'm assuming that changes in the `upgradedProperties` file would be enough to invalidate whatever cache we are trying to invalidate here, no?

Reply:
In first commits I actually just hashed a file. But then I figured this can create a different hash unintentionally, e.g. if json stays the same, but order of nodes in it changes for some reason. 

So now I calculate hash for every property and then I calculate combined stable hash in this `InstrumentedMetadataMergeTask`.